lib/Common/Memory/WriteBarrierMacros.h

```diff
@@ -17,6 +17,8 @@
 
 #define FieldWithBarrier(type, ...) \
     typename WriteBarrierFieldTypeTraits<type, ##__VA_ARGS__>::Type
+#define FieldWithBarrierT(type, ...) \
```

Comment:
Declarations within template classes with dependent name need `typename`. 
```  template<class T>
    class InlineCachePointerArray
    {
    public:
        FieldWithBarrier(Field(T*)*) inlineCaches;
   ...
   }
```
Whereas the uses in `FunctionBody` structures should not have `typename`.